pydomcfg/domzgr/zco.py

```diff
@@ -0,0 +1,274 @@
+        ppa2: float = 0.0,
+        ppkth2: float = 0.0,
+        ppacr2: float = 0.0,
+        ln_e3_dep: bool = True,
+    ) -> Dataset:
+        """
+        Generate NEMO geopotential z-coordinates model levels.
+
+        Parameters
+        ----------
+        ppdzmin: float
+            Minimum thickness for the top layer (units: m)
+        pphmax: float
+            Depth of the last W-level (total depth of the ocean basin) (units: m)
+        ppacr: int
+            Stretching factor, nondimensional > 0. The larger ppacr,
+            the smaller the stretching. Values from 3 to 10 are usual.
+            (default = 0., no stretching, uniform grid)
+        ppkth: float
+            Model level at which approximately max stretching occurs.
+            Nondimensional > 0, usually of order 1/2 or 2/3 of jpk.
+            (default = 0., i.e. no stretching, uniform grid)
+        ppsur: float, optional
+            Coeff. controlling distibution of vertical levels
+            (default = None, i.e. computed from ppdzmin, pphmax, ppkth and ppacr)
+        ppa0: float, optional
+            Coeff. controlling distibution of vertical levels
+            (default = None, i.e. computed from ppdzmin, pphmax, ppkth and ppacr)
+        ppa1: float, optional
+            Coeff. controlling distibution of vertical levels
+            (default = None, i.e. computed from ppdzmin, pphmax, ppkth and ppacr)
+        ldbletanh: bool
+            Logical flag to use or not double tanh stretching function
+            (default = False)
+        ppa2: float
+            Double tanh stretching function parameter
+        ppkth2: float
+            Double tanh stretching function parameter
+        ppacr2: float
+            Double tanh stretching function parameter
+        ln_e3_dep: bool
+            Logical flag to comp. e3 as fin. diff. (True) or
+            analyt. (False) (default = True)
+
+        Returns
+        -------
+        Dataset
+           Describing the 3D geometry of the model
+
+        Raises
+        -------
+        ValueError
+            If ldbletanh = True and parametrs are equal to 0.
+
+        """
+        self._ppdzmin = ppdzmin
+        self._pphmax = pphmax
+        self._ppkth = ppkth
+        self._ppacr = ppacr
+        self._ldbletanh = ldbletanh
+        self._ppa2 = ppa2
+        self._ppkth2 = ppkth2
+        self._ppacr2 = ppacr2
+        self._is_uniform = not ppkth or not ppacr
+        self._ln_e3_dep = ln_e3_dep
+
+        # Checking consistency of input parameters
+        if self._ldbletanh:
+            if self._ppa2 * self._ppkth2 * self._ppacr2 == 0.0:
+                raise ValueError(
+                    "ppa2, ppkth2 and ppacr2 MUST be > 0. " "when ldbletanh = True"
+                )
+
+        ds = self._init_ds()
+
+        # computing coeff. if needed
+        if not self._is_uniform:
+            self._ppsur, self._ppa0, self._ppa1 = self._compute_pp(ppsur, ppa0, ppa1)
```

Comment:
I think we are missing an error here, i.e. the case where is uniform and the user provides any of `[ppsur, ppa0, ppa1]`

Reply:
If the user wants a uniform grid (acr or kth == 0) then we don't need ppsur, ppa0 and ppa1 have - they are not used. So even if the user provide ppsur, ppa0 and ppa1, the condition of uniform grid prevent to set them. I think we are safe like this.

Reply:
OK! Still the user is probably misunderstanding something... Let's add a warning?

Reply:
OK! Still the user is probably misunderstanding something... Let's add a warning?

Reply:
Fine by me